types/context_cache.go

```diff
@@ -37,73 +31,53 @@ func (c *ContextMemCache) UpsertDeferredSends(moduleAccount string, amount Coins
 	}
-
-	return c.deferredSends.SafeSub(moduleAccount, amount)
-}
-
-func (c *ContextMemCache) RangeOnDeferredSendsAndDelete(apply func(recipient string, amount Coins)) {
-	c.deferredSends.RangeOnMapping(apply)
-}
-
-func (c *ContextMemCache) UpsertDeferredWithdrawals(moduleAccount string, amount Coins) error {
-	if !amount.IsValid() {
-		return sdkerrors.Wrap(sdkerrors.ErrInvalidCoins, amount.String())
-	}
-
-	// Separate locks needed for all mappings - atmoic transaction needed
 	c.deferredBankOpsLock.Lock()
 	defer c.deferredBankOpsLock.Unlock()
 
-	// If there's already a pending deposit then subtract it from that amount first
-	// or else add it to the deferredWithdrawals mapping
-	ok := c.deferredSends.SafeSub(moduleAccount, amount)
+	originalBalance, ok := c.deferredSends.Get(moduleAccount)
 	if !ok {
-		c.deferredWithdrawals.UpsertMapping(moduleAccount, amount)
+		// run the remainder handler on the full amount
+		err := remainderHandler(amount)
+		// no need for revert because we dont even attempt a saturating sub, bubble up error if applicable
+		return err
 	}
-	return nil
+	// found a balance, try to perform the logic
+	remainder := c.deferredSends.SaturatingSub(moduleAccount, amount)
+	if remainder.IsZero() {
+		// no remainder, return now
+		return nil
+	}
+	err := remainderHandler(remainder)
+	if err != nil {
+		// revert the map subtraction and the bubble up error
+		c.deferredSends.Set(moduleAccount, originalBalance)
+	}
+	return err
 }
 
-// This inserts or updates an entry for a module account for a deferred withdrawal.
-// This should be performed AFTER checking for a sufficient balance in the underlying bank balances for that module account.
-// Additionally, this does not attempt to do any safe subtraction from pending sends, so if that behavior is preferred, it will need to be checked separately.
-func (c *ContextMemCache) UpsertDeferredWithdrawalsNoSafeSub(moduleAccount string, amount Coins) error {
+func (c *ContextMemCache) SafeSubDeferredSends(moduleAccount string, amount Coins) bool {
 	if !amount.IsValid() {
-		return sdkerrors.Wrap(sdkerrors.ErrInvalidCoins, amount.String())
+		panic(sdkerrors.Wrap(sdkerrors.ErrInvalidCoins, amount.String()))
 	}
 
-	// Separate locks needed for all mappings - atmoic transaction needed
-	c.deferredBankOpsLock.Lock()
-	defer c.deferredBankOpsLock.Unlock()
-
-	c.deferredWithdrawals.UpsertMapping(moduleAccount, amount)
-	return nil
-}
-
-func (c *ContextMemCache) RangeOnDeferredWithdrawalsAndDelete(apply func(recipient string, amount Coins)) {
-	c.deferredWithdrawals.RangeOnMapping(apply)
+	return c.deferredSends.SafeSub(moduleAccount, amount)
 }
 
-func (c *ContextMemCache) ApplyOnAllDeferredOperationsAndDelete(apply func(recipient string, amount Coins)) {
-	c.RangeOnDeferredSendsAndDelete(apply)
-	c.RangeOnDeferredWithdrawalsAndDelete(apply)
+func (c *ContextMemCache) RangeOnDeferredSendsAndDelete(apply func(recipient string, amount Coins)) {
+	c.deferredSends.RangeAndRemove(apply)
```

Comment:
This probably doesn't since it's supposed to be called a the end synchronously but could we add a lock or add a comment that it shouldn't be ran concurrently?